pl_examples/hydra_examples/conf/optimizer.py

```diff
@@ -0,0 +1,145 @@
+from dataclasses import dataclass
+
+from hydra.core.config_store import ConfigStore
+from hydra.types import ObjectConf
+
+
+@dataclass
+class AdamConf:
+    betas: tuple = (0.9, 0.999)
+    lr: float = 1e-3
+    eps: float = 1e-8
+    weight_decay: float = 0
+    amsgrad: bool = False
+
+
+@dataclass
+class Adam(ObjectConf):
+    cls: str = "torch.optim.Adam"
+    params: AdamConf = AdamConf()
+
+
+@dataclass
+class AdamaxConf:
+    betas: tuple = (0.9, 0.999)
+    lr: float = 1e-3
+    eps: float = 1e-8
+    weight_decay: float = 0
+
+
+@dataclass
+class Adamax(ObjectConf):
+    cls: str = "torch.optim.Adamax"
+    params: AdamConf = AdamConf()
+
+
+@dataclass
+class AdamW(ObjectConf):
+    cls: str = "torch.optim.AdamW"
+    params: AdamConf = AdamConf()
+
+
+@dataclass
+class ASGDConf:
+    alpha: float = 0.75
+    lr: float = 1e-3
+    lambd: float = 1e-4
+    t0: int = 1e6
+    weight_decay: float = 0
+
+
+@dataclass
+class ASGD(ObjectConf):
+    cls: str = "torch.optim.ASGD"
+    params: ASGDConf = ASGDConf()
+
+
+@dataclass
+class LBFGSConf:
+    lr: float = 1
+    max_iter: int = 20
+    max_eval: int = 25
+    tolerance_grad: float = 1e-5
+    tolerance_change: float = 1e-9
+    history_size: int = 100
+    line_search_fn: str = None
+
+
+@dataclass
+class LBFGS(ObjectConf):
+    cls: str = "torch.optim.LBFGS"
+    params: LBFGSConf = LBFGSConf()
+
+
+@dataclass
+class RMSpropConf:
+    lr: float = 1e-2
+    momentum: float = 0
+    alpha: float = 0.99
+    eps: float = 1e-8
+    centered: bool = True
+    weight_decay: float = 0
+
+
+@dataclass
+class RMSprop(ObjectConf):
+    cls: str = "torch.optim.RMSprop"
+    params: RMSpropConf = RMSpropConf()
+
+
+@dataclass
+class RpropConf:
+    lr: float = 1e-2
+    momentum: tuple = (0.5, 1.2)
+    step_sizes: tuple = (1e-6, 50)
+
+
+@dataclass
+class Rprop(ObjectConf):
+    cls: str = "torch.optim.Rprop"
+    params: RpropConf = RpropConf()
+
+
+@dataclass
+class SGDConf:
+    lr: float = 1e-2
+    momentum: float = 0
+    weight_decay: float = 0
+    dampening: float = 0
+    nesterov: bool = False
+
+
+@dataclass
+class SGD(ObjectConf):
+    cls: str = "torch.optim.SGD"
+    params: SGDConf = SGDConf()
+
+
+ConfigStore.instance().store(
+    group="opt", name="adam", node=Adam,
+)
+
+ConfigStore.instance().store(
+    group="opt", name="adamax", node=Adamax,
+)
+
+ConfigStore.instance().store(
+    group="opt", name="adamw", node=AdamW,
+)
+
+ConfigStore.instance().store(
+    group="opt", name="asgd", node=ASGD,
+)
+
+ConfigStore.instance().store(
+    group="opt", name="lbfgs", node=LBFGS,
+)
+ConfigStore.instance().store(
+    group="opt", name="rmsprop", node=RMSprop,
+)
+ConfigStore.instance().store(
+    group="opt", name="rprop", node=Rprop,
+)
+ConfigStore.instance().store(
+    group="opt", name="sgd", node=SGD,
+)
```

Comment:
Copying an idea from the chat.
Right now you have:

```python
@dataclass
class SGDConf:
    lr: float = 1e-2
    momentum: float = 0
    weight_decay: float = 0
    dampening: float = 0
    nesterov: bool = False


@dataclass
class SGD(ObjectConf):
    cls: str = "torch.optim.SGD"
    params: SGDConf = SGDConf()

ConfigStore.instance().store(
    group="opt", name="sgd", node=SGD,
)
```

I think this will also work:
```python
@dataclass
class SGDConf:
    lr: float = 1e-2
    momentum: float = 0
    weight_decay: float = 0
    dampening: float = 0
    nesterov: bool = False

ConfigStore.instance().store(
    group="opt", name="sgd", node=ObjectConf(cls="torch.optim.SGD", params=SGDConf())
)
```

Reply:
Great. assuming it works as expected?

Reply:
I like doing the ConfigStore registration for each component in a different file. Makes this much cleaner. Well done! :)

Reply:
Yes, this works very nicely!